pkg/scripts/install.go

```diff
@@ -0,0 +1,184 @@
+	return i.run(scripts, dstDir)
+}
+
+type downloadResult struct {
+	location string
+	err      error
+}
+
+func (i *installer) run(scripts []api.Script, dstDir string) []api.InstallResult {
+	if len(scripts) == 0 {
+		return []api.InstallResult{}
+	}
+
+	var userResults, sourceResults, defaultResults map[api.Script]*downloadResult
+
+	// get scripts from user provided URL
+	if i.scriptsURL != "" {
+		userResults = i.download(i.scriptsURL, scripts, filepath.Join(dstDir, api.UserScripts))
+	}
+
+	// get scripts from source
+	sourceResults = make(map[api.Script]*downloadResult)
+	for _, script := range scripts {
+		sourceResults[script] = &downloadResult{location: api.SourceScripts}
+		file := filepath.Join(dstDir, api.SourceScripts, string(script))
+		if !i.fs.Exists(file) {
+			sourceResults[script].err = errors.NewDownloadError(file, -1)
+		}
+	}
+
+	// get scripts from default URL
+	defaultURL, err := i.docker.GetScriptsURL(i.image)
+	if err == nil && defaultURL != "" {
+		defaultResults = i.download(defaultURL, scripts, filepath.Join(dstDir, api.DefaultScripts))
+	} else {
+		glog.Warningf("Image %s does not contain STI_SCRIPTS_URL defined!", i.image)
+	}
+
+	return i.install(scripts, userResults, sourceResults, defaultResults, dstDir)
+}
+
+func (i *installer) download(scriptsURL string, scripts []api.Script, dstDir string) map[api.Script]*downloadResult {
+	result := make(map[api.Script]*downloadResult)
+	for _, script := range scripts {
+		result[script] = &downloadResult{location: scriptsURL}
+		url, err := url.Parse(scriptsURL + "/" + string(script))
+		if err != nil {
+			result[script].err = err
+			continue
+		}
+		result[script].err = i.downloader.Download(url, filepath.Join(dstDir, string(script)))
+	}
+
+	return result
+}
+
+func (i *installer) install(scripts []api.Script, userResults, sourceResults, defaultResults map[api.Script]*downloadResult, dstDir string) []api.InstallResult {
+	resultList := []api.InstallResult{}
+
+	locationsResultsMap := map[string]map[api.Script]*downloadResult{
+		api.UserScripts:    userResults,
+		api.SourceScripts:  sourceResults,
+		api.DefaultScripts: defaultResults,
+	}
+
+	for _, script := range scripts {
+		result := api.InstallResult{Script: script}
+
+		for _, location := range locationsOrder {
+			locationResults, ok := locationsResultsMap[location]
+			if !ok || locationResults == nil {
+				continue
+			}
+			downloadResult, ok := locationResults[script]
+			if !ok {
+				continue
+			}
+			result.URL = downloadResult.location
+			// if location results are erroneous we store error in result object
+			// and continue searching other locations
+			if downloadResult.err != nil {
+				// one exception is when error contains information about scripts being inside the image
+				if e, ok := downloadResult.err.(errors.Error); ok && e.ErrorCode == errors.ErrScriptsInsideImage {
+					// in which case update result object and break further searching
+					result.Error = nil
```

Comment:
It would complicate code a little bit, since I'm always returning the last error for the user as an information.